qlib/TableMapper.qm

```diff
@@ -637,65 +669,64 @@ on_error table_mapper.rollback();
-                    *hash th = insertRow(row);
+                    # FIXME --PQ 21-Mar-2017
+                    # for const hashes, we have a problem here, since
+                    # the row is actually inserted multiple times into the table,
+                    # but appears only once on the output!
+                    *hash th = insertRow(row + crec);
                     if (th) {
-                        if (h)
-                            map h.$1 += th.$1, th.keyIterator();
-                        else
-                            h = th;
+                        hash bulkth = record2Bulk(th);
+                        addBulkToBulk(\h, bulkth);
                     }
                 }
                 return h;
             }
 
-            hash dh;
-
-            # first copy all 1:1 mappings to the output hash
-            if (identl)
-                dh = rec{identl};
+            if (rec_list_size > 0) {
+                hash dh;
 
-            # copy all runtime mappings to the output hash
-            map dh{$1.key} = m_runtime{$1.value}, rconsth.pairIterator();
+                # identl and rconsth entries are added to 'dh' and not in 'hbuf'
+                # directly since identl and rconsth may change at each call of
+                # queueData() (unlike consth).
 
-            int size = rec.firstValue().size();
-            map mapFieldIntern(\dh, $1, rec, size), mapd.keyIterator();
-            count += size;
+                # first copy all 1:1 mappings to the output hash
+                if (identl)
+                    dh = rec{identl};
 
-            # map record data to get the keys for the buffer
-            if (!hbuf) {
-                # copy all constant mappings to the output hash
-                if (consth)
-                    hbuf += consth;
-                if (crec)
-                    hbuf += crec;
-
-                # add mapped data
-                hbuf += dh;
-            }
-            else
-                map hbuf{$1.key} += $1.value, dh.pairIterator();
+                # copy all runtime mappings to the output hash
+                map dh{$1.key} = m_runtime{$1.value}, rconsth.pairIterator();
 
-            # return nothing if nothing needs to be flushed
-            if (hbuf.firstValue().size() < insert_block)
-                return;
+                map mapFieldIntern(\dh, $1, rec, True, rec_list_size, crec), mapd.keyIterator();
+                count += rec_list_size;
 
-            hash h;
-            # return all target data
-            *hash th = flushIntern();
+                # map record data to get the keys for the buffer
+                if (!hbuf) {
+                    # copy all constant mappings to the output hash
+                    # consth, rconsth, identh and crec are removed from mapd,
+                    # so they must be re-added now.
+                    # FIXME what is not nice is that these fields are also
```

Comment:
`consth` is not processed in `Mapper::mapFieldIntern()`; it is removed from `mapd`; it is only processed in `Mapper::mapDataIntern()` and therefore must be processed by any high-level method that calls `Mapper::mapFieldIntern()`